File: MonotonicStack/316_Remove_Duplicate_Letters/316_Remove_Duplicate_Letters.cpp
// Note: This question is the same as 1081: https://leetcode.com/problems/smallest-subsequence-of-distinct-characters/
class Solution {
public:
    string removeDuplicateLetters(string s) {
        
        unordered_map<char, unsigned int> count;
        for(auto& c: s)
            count[c]++;
    
        stack<char> stack;
        set<char> in_stack;

        for(auto& c: s) {
            
            count[c]--;
            
            if(is_in_stack(c, stack))
                continue;

            while(!stack.empty() && c<stack.top() && count[stack.top()]>0) {
                in_stack.erase(stack.top());
                stack.pop();
            }

            stack.push(c);
            in_stack.insert(c);

        }

    	string res;
	    while(!stack.empty()) {
		    res.push_back(stack.top());
		    stack.pop();
	    }
        reverse(res.begin(), res.end());

        return res;
    
    }

    bool is_in_stack(char c, stack<char> stack) {
        while (!stack.empty()) {
            if (stack.top() == c) return true;
            stack.pop();
        }
        return false;
    }


};

File: MonotonicStack/316_Remove_Duplicate_Letters/316_Remove_Duplicate_Letters.java
// Note: This question is the same as 1081: https://leetcode.com/problems/smallest-subsequence-of-distinct-characters/
class Solution {
    public String removeDuplicateLetters(String s) {

        char[] ss = s.toCharArray();

        HashMap<Character, Integer> count = new HashMap<Character, Integer>();
        for (char c : ss) {
            if(count.containsKey(c)) 
                count.put(c, count.get(c) + 1);
            else 
                count.put(c, 1);            
        }

        Stack<Character> stack = new Stack<>();
        Set<Character> in_stack = new HashSet<>(); 

        for(Character c: ss) {
            
            count.put(c, count.get(c) - 1);

            if(stack.search(c) != -1)
                continue;

            while(!stack.empty() && c<stack.peek() && count.get(stack.peek())>0)
                in_stack.remove(stack.pop());
            
            stack.push(c);
            in_stack.add(c);

        }

        StringBuilder res = new StringBuilder();

	    while(!stack.empty())
		    res.append(stack.pop());

        res.reverse();

        return res.toString();
    }
}
